pkg/drainer/drainer.go

```diff
@@ -39,6 +39,7 @@ type Options struct {
 	DeleteLocalData     bool
 	Selector            string
 	PodSelector         string
+	DrainDelay          time.Duration `mapstructure:"drain-delay"`
```

Comment:
sorry, I don't understand the significance of the `mapstructure:"drain-delay"` part, the code is not fresh in my mind like it used to be, but AFAIR there is no serialisation here, happy to learn more

Reply:
In [cmd/kubedrainer/serve.go#181](https://github.com/VirtusLab/kubedrainer/blob/84492c50d88a1a2c70c5edf6f425add5fe4cf716/cmd/kubedrainer/serve.go#L181) you call settings.Parse() which calls Unmarshall with mapstructure decoder:

```go
err := viper.Unmarshal(target, func(config *mapstructure.DecoderConfig) {
	config.ErrorUnused = false
	config.ZeroFields = false
})
```

I tested it without a mapstructure tag and it couldn't parse flags properly: the value of `DrainDelay` stayed default - 0s.

Reply:
In fact, I believe, that some other flags, like for example `pod-selector` cannot be parsed properly as well:

I added simple debug output at the end of serve/Parse() function to check it:

```go
// Parse parses all flags and settings to options
func (o *ServeOptions) Parse(cmd *cobra.Command) error {
	settings.Bind(cmd.Flags()) // needs to be run inside the command and before any viper usage for flags to be visible

	if debug {
		log.Debug().Msgf("All keys: %+v", viper.AllKeys())
		log.Debug().Msgf("All settings: %+v", viper.AllSettings())
		cmd.Flags().VisitAll(func(flag *pflag.Flag) {
			log.Debug().Msgf("'%s' -> flag: '%+v' | setting: '%+v'", flag.Name, flag.Value, viper.Get(flag.Name))
		})
		log.Debug().Msgf("Settings: %+v", *o)
	}
	if err := settings.Parse(o.Kubernetes); err != nil {
		return err
	}
	if err := settings.Parse(o.Drainer); err != nil {
		return err
	}
	if err := settings.Parse(o.AWS); err != nil {
		return err
	}
	log.Debug().Msgf("Parsed settings: %+v", *o) # <--- this one
	return nil
}
```

and I see this behaviour when starting kubedrainer with `./kubedrainer -d serve --pod-selector checking`:

- without mapstructure tag for `PodSelector`:

```json
{"level":"debug","time":1613463410,"message":"Settings: {Kubernetes:{ConfigFlags:{CacheDir:/Users/rtim/.kube/http-cache KubeConfig: ClusterName: AuthInfoName: Context: Namespace:<nil> APIServer: Insecure:false CertFile: KeyFile: CAFile: BearerToken: Impersonate:<nil> ImpersonateGroup:<nil> Username:<nil> Password:<nil> Timeout:0 clientConfig:<nil> lock:{state:0 sema:0} usePersistentConfig:true}} Drainer:{Node: Force:false GracePeriodSeconds:-1 IgnoreAllDaemonSets:true Timeout:1m0s DeleteLocalData:true Selector: PodSelector: DrainDelay:0s} AWS:0xc00003a6e0}"}

{"level":"debug","time":1613463410,"message":"Parsed settings: {Kubernetes:{ConfigFlags:{CacheDir:/Users/rtim/.kube/http-cache KubeConfig: ClusterName: AuthInfoName: Context: Namespace:<nil> APIServer: Insecure:false CertFile: KeyFile: CAFile: BearerToken: Impersonate:<nil> ImpersonateGroup:<nil> Username:<nil> Password:<nil> Timeout:0 clientConfig:<nil> lock:{state:0 sema:0} usePersistentConfig:true}} Drainer:{Node: Force:false GracePeriodSeconds:-1 IgnoreAllDaemonSets:true Timeout:1m0s DeleteLocalData:true Selector: PodSelector: DrainDelay:0s} AWS:0xc00003a6e0}"}
```
you can see that PodSelector(at the end of JSONs) doesn't change

- with mapstructure tag `mapstructure:"pod-selector"`:

```json
{"level":"debug","time":1613463080,"message":"Settings: {Kubernetes:{ConfigFlags:{CacheDir:/Users/rtim/.kube/http-cache KubeConfig: ClusterName: AuthInfoName: Context: Namespace:<nil> APIServer: Insecure:false CertFile: KeyFile: CAFile: BearerToken: Impersonate:<nil> ImpersonateGroup:<nil> Username:<nil> Password:<nil> Timeout:0 clientConfig:<nil> lock:{state:0 sema:0} usePersistentConfig:true}} Drainer:{Node: Force:false GracePeriodSeconds:-1 IgnoreAllDaemonSets:true Timeout:1m0s DeleteLocalData:true Selector: PodSelector: DrainDelay:0s} AWS:0xc00029e5f0}"}

{"level":"debug","time":1613463080,"message":"Parsed settings: {Kubernetes:{ConfigFlags:{CacheDir:/Users/rtim/.kube/http-cache KubeConfig: ClusterName: AuthInfoName: Context: Namespace:<nil> APIServer: Insecure:false CertFile: KeyFile: CAFile: BearerToken: Impersonate:<nil> ImpersonateGroup:<nil> Username:<nil> Password:<nil> Timeout:0 clientConfig:<nil> lock:{state:0 sema:0} usePersistentConfig:true}} Drainer:{Node: Force:false GracePeriodSeconds:-1 IgnoreAllDaemonSets:true Timeout:1m0s DeleteLocalData:true Selector: PodSelector:checking DrainDelay:0s} AWS:0xc00029e5f0}"}
```

Reply:
you're absolutely right, thank you for looking at other fields as well

Reply:
Do you want to me to continue in this PR or should we merge this one and open a new PR to fix other fields?

Reply:
I'll merge this one, and if you would be son kind we'll gladly accept the fix for the other fields (was about to start working on it) :)

Reply:
sure, I can fix other fields